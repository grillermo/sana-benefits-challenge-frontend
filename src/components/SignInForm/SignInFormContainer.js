import { useContext } from 'react'

import { useLocation } from "wouter";

import { signIn } from '../../utils/apiClient'
import SignInForm from './SignInForm'
import AuthContext from '../../contextProviders/authContext'
import emailValidation from '../../utils/emailValidation'

export function handleSubmit(values, setSubmitting, authDispatch, setLocation) {
  setSubmitting(true)

  return signIn(values).then((user) => {
    authDispatch({
      type: 'SIGNIN',
      payload: user
    })

    setLocation('/configuration')
  }).catch((errorMessage) => {
    setSubmitting(false)
  })
}

const SignInFormContainer = () => {
  const { dispatch: authDispatch } = useContext(AuthContext);
  const setLocation = useLocation()[1];

  return (
    <SignInForm
      emailValidation={emailValidation}
      handleSubmit={(values, { setSubmitting }) => (
        handleSubmit(values, setSubmitting, authDispatch, setLocation)
      )}
    />
  )
};

export default SignInFormContainer
