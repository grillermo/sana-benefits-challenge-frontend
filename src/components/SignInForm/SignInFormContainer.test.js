import React from 'react'
import { mount } from 'enzyme'
import axios from 'axios'

import SignInFormContainer, { handleSubmit } from './SignInFormContainer'
import AuthContext from '../../contextProviders/authContext'

jest.mock("axios")

describe('SignInFormContainer', () => {
  it('matches the snapshot', async () => {
    const signInFormContainer = mount(
      <AuthContext.Provider value={{
        dispatch: jest.fn()
      }} >
        <SignInFormContainer />
      </AuthContext.Provider>
    )

    expect(signInFormContainer).toMatchSnapshot();
  });
})
  
describe('handleSubmit', () => {
  // arguments
  const values = {}
  const setSubmitting = jest.fn()
  const authDispatch = jest.fn()
  const setLocation = jest.fn()
  const user = { id: 1, email: "email@domain.com" }
  const token = 'eyJhbGciOiJIUzI1...'

  beforeEach(() => {
    jest.clearAllMocks()

    let mockedResponse = {
      data: user,
      status: 200,
      headers: {
        authorization: 'Bearer '+token
      },
    }

    axios.post.mockImplementationOnce(() => Promise.resolve(mockedResponse))
  })

  it('sets the submitting flag', async () => {
    await handleSubmit(values, setSubmitting, authDispatch, setLocation)

    expect(setSubmitting).toHaveBeenCalledWith(true);
  })

  it('signs in the user', async () => {
    await handleSubmit(values, setSubmitting, authDispatch, setLocation)

    expect(authDispatch).toHaveBeenCalledWith({
      type: 'SIGNIN',
      payload: {
        token: token,
        user: user,
      }
    })
  })

  it('redirecs the user to configuration', async () => {
    await handleSubmit(values, setSubmitting, authDispatch, setLocation)

    expect(setLocation).toHaveBeenCalledWith('/configuration')
  })
})

