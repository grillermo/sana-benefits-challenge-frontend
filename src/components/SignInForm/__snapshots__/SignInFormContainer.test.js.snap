// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignInFormContainer matches the snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <label>
    email:
    <input
      autoComplete="username"
      name="email"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="joe@google.com"
      type="email"
      value=""
    />
  </label>
  <label>
    password:
    <input
      autoComplete="current-password"
      name="password"
      onBlur={[Function]}
      onChange={[Function]}
      placeholder="*******"
      type="password"
      value=""
    />
  </label>
  <button
    disabled={false}
    type="submit"
  >
    Submit
  </button>
</form>
`;
